src/pymor/algorithms/bernoulli.py

```diff
@@ -0,0 +1,173 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2020 pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.core.logger import getLogger
+from pymor.operators.constructions import IdentityOperator
+
+
+def solve_bernoulli(A, E, B, trans=False, maxiter=100, after_maxiter=3, tol=1e-8):
+    """Compute a solution factor of a Bernoulli equation.
+
+    Returns a matrix :math:`Y` such that :math:`X = Y Y^H`is an approximate solution
+    of a (generalized) algebraic Bernoulli equation:
+
+    - if trans is `True`
+
+      .. math::
+          A^H X E + E^H X A
+          - E^H X B B^H X E = 0.
+
+    - if trans is `False`
+
+      .. math::
+          A X E^H + E X A^H
+          - E X B^H B X E^H = 0.
+
+    This function is based on [BBQ07]_.
+
+    Parameters
+    ----------
+    A
+        The operator A as a 2D |NumPy array|.
+    E
+        The operator E as a 2D |NumPy array| or `None`.
+    B
+        The operator B as a 2D |NumPy array|.
+    trans
+        Wether to solve transposed or standard Bernoulli equation.
+    maxiter
+        The maximum amount of iterations.
+    after_maxiter
+        The number of iterations which are to be performed after tolerance is reached.
+    tol
+        Tolerance for stopping criterion based on relative change of iterates.
+
+    Returns
+    -------
+    Y
+        The solution factor as a |NumPy array|.
+    """
+
+    logger = getLogger('pymor.algorithms.bernoulli.solve_bernoulli')
+
+    n = len(A)
+    after_iter = 0
+
+    assert n != 0
+
+    if E is None:
+        E = np.eye(n)
+
+    if not trans:
+        E = E.conj().T
+        A = A.conj().T
+        B = B.conj().T
+
+    for i in range(maxiter):
+        Aprev = A
+        lu, piv = spla.lu_factor(A.conj().T)
+        if E is not None:
+            detE = spla.det(E)
+        else:
+            detE = 1
+        if detE.real > 0.:
+            c = np.abs(np.prod(np.diag(lu)) / detE)**(1. / n)  # |det(A_k)/det(E)|^(1/n)
```

Comment:
I'm worried that the determinant computation might overflow or underflow. Maybe use the logarithm? NumPy has [`slogdet`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.slogdet.html#numpy.linalg.slogdet).